pkg/kinflate/resource/configmap.go

```diff
@@ -39,29 +44,83 @@ func NewFromConfigMap(cm manifest.ConfigMap) (*Resource, error) {
 	return &Resource{Data: data}, nil
 }
 
-func makeConfigMap(cm manifest.ConfigMap) (*corev1.ConfigMap, error) {
+func makeConfigMap(cm manifest.ConfigMap, l loader.Loader) (*corev1.ConfigMap, error) {
 	corev1cm := &corev1.ConfigMap{}
 	corev1cm.APIVersion = "v1"
 	corev1cm.Kind = "ConfigMap"
 	corev1cm.Name = cm.Name
 	corev1cm.Data = map[string]string{}
 
-	// TODO: move the configmap helpers functions in this file/package
-	if cm.EnvSource != "" {
-		if err := cutil.HandleConfigMapFromEnvFileSource(corev1cm, cm.EnvSource); err != nil {
-			return nil, err
+	envPairs, err := keyValuesFromEnvFile(l, cm.EnvSource)
+	if err != nil {
+		return nil, fmt.Errorf("error reading keys from env source file: %s %v", cm.EnvSource, err)
+	}
+
+	literalPairs, err := keyValuesFromLiteralSources(cm.LiteralSources)
+	if err != nil {
+		return nil, fmt.Errorf("error reading key values from literal sources: %v", err)
+	}
+
+	filePairs, err := keyValuesFromFileSources(l, cm.FileSources)
+	if err != nil {
+		return nil, fmt.Errorf("error reading key values from file sources: %v", err)
+	}
+
+	allPairs := append(append(envPairs, literalPairs...), filePairs...)
```

Comment:
Instead of appending everywhere and then check for errors. I'm wondering if we should have a sort of class that tracks the key (make sure they are valid as they are created), and fails when there is a duplicate? Instead of maintaining a list, we could use a `map[string]string` (key -> value). wdyt?

Reply:
updated the code here (we also looked at it offline).